utilities/workers/worker_manager.py

```diff
@@ -195,3 +218,35 @@ def join_workers(self) -> None:
         """
         for worker in self.__workers:
             worker.join()
+
+    def check_and_restart_dead_workers(self) -> None:
+        """
+        Check and restart dead workers.
+        """
+        for worker in self.__workers:
+            if not worker.is_alive():
+                # Log the error
+                frame = inspect.currentframe()
+                worker_name_string = self.__worker_properties.get_worker_name() + " worker"
+                self.__local_logger.error("Worker died, restarting " + worker_name_string, frame)
+
+                # Terminate and join worker
+                worker.terminate()
+                worker.join()
+                del worker
+                worker = None
+
+                # Drain the preceding queues
+                input_queues = self.__worker_properties.get_input_queues()
+                for queue in input_queues:
+                    queue.drain_queue()
```

Comment:
It is inefficient to drain the queues every time the loop encounters a dead worker. Pull this out of the loop so that it only needs to be done once.